File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.java
class WordDictionary {
    private static class TrieNode {
        private boolean isEnd;
        private TrieNode[] children = new TrieNode[26];

        public TrieNode() {
            isEnd = false;
        }
    }
    
    private final TrieNode root;
    public WordDictionary() {
        root = new TrieNode();
    }
    
    public void addWord(String word) {
        TrieNode node = root;

        for (char c : word.toCharArray()) {
            if (node.children[c - 'a'] == null) {
                node.children[c - 'a'] = new TrieNode();
            }
            
            node = node.children[c - 'a'];
        }
        node.isEnd = true;
    }
    
    public boolean search(String word) {
        return search(word, root, 0);
    }

    private boolean search(String word, TrieNode node, int idx) {
        if (idx == word.length()) {
            return node.isEnd;
        }

        char c = word.charAt(idx);

        if (c == '.') {
            for (int i = 0; i < 26; i++) {
                if (node.children[i] != null && search(word, node.children[i], idx+1)) {
                    return true;
                }
            }
            return false;
        } else {
            if (node.children[c - 'a'] == null) {
                return false;
            } else {
                return search(word, node.children[c - 'a'], idx + 1);
            }
        }
    }

    
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.py
class Node:
    def __init__(self):
        self.children = {}
        self.end = False
class WordDictionary:

    def __init__(self):
        self.root = Node()
        
    def addWord(self, word: str) -> None:
        curr = self.root

        for c in word:
            if not c in curr.children:
                curr.children[c] = Node()
            
            curr = curr.children[c]

        curr.end = True

    def search(self, word: str) -> bool:
        def dfs(node, pos):
            if pos == len(word):
                return node.end

            char = word[pos]

            if char == ".":
                return any(dfs(child, pos + 1) for child in node.children.values())
            elif char in node.children:
                return dfs(node.children[char], pos + 1)

            return False

        return dfs(self.root, 0)
        # stack = [(self.root, word)]

        # while stack:
        #     currNode, currWord = stack.pop()

        #     if not currWord:
        #         if currNode.end:
        #             return True
        #         continue
            
        #     if currWord[0] == ".":
        #         for nextNode in currNode.children.values():
        #             stack.append((nextNode, currWord[1:]))
        #     else:
        #         if currWord[0] in currNode.children:
        #             stack.append((currNode.children[currWord[0]], currWord[1:]))

        # return False
        




# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)